osf/management/commands/update_institution_sso_email_domain.py

```diff
@@ -0,0 +1,146 @@
+        if not institution:
+            message = 'Error: invalid institution ID [{}]'.format(institution_id)
+            logger.error(message)
+            sentry.log_message(message)
+            return
+
+        # Find all users affiliated with the given institution
+        affiliated_users = OSFUser.objects.filter(affiliated_institutions___id=institution_id)
+
+        # Update email domain for each user
+        update_result = {
+            UpdateResult.SUCCEEDED.name: {},
+            UpdateResult.SKIPPED.name: {},
+            UpdateResult.FAILED.name: {},
+            UpdateResult.ERRORED.name: {},
+        }
+        for user in affiliated_users:
+            user_result = update_email_domain(user, src_domain, dst_domain, dry_run=dry_run)
+            if user_result.get(UpdateResult.SUCCEEDED.name):
+                update_result.get(UpdateResult.SUCCEEDED.name)[user._id] = user_result.get(UpdateResult.SUCCEEDED.name)
+            if user_result.get(UpdateResult.SKIPPED.name):
+                update_result.get(UpdateResult.SKIPPED.name)[user._id] = user_result.get(UpdateResult.SKIPPED.name)
+            if user_result.get(UpdateResult.FAILED.name):
+                update_result.get(UpdateResult.FAILED.name)[user._id] = user_result.get(UpdateResult.FAILED.name)
+            if user_result.get(UpdateResult.ERRORED.name):
+                update_result.get(UpdateResult.ERRORED.name)[user._id] = user_result.get(UpdateResult.ERRORED.name)
+
+        # Output update results to console
+        logger.info(f'{UpdateResult.SUCCEEDED.name} = {update_result.get(UpdateResult.SUCCEEDED.name)}')
+        logger.info(f'{UpdateResult.SKIPPED.name} = {update_result.get(UpdateResult.SKIPPED.name)}')
+        logger.warning(f'{UpdateResult.FAILED.name} = {update_result.get(UpdateResult.FAILED.name)}')
+        logger.error(f'{UpdateResult.ERRORED.name} = {update_result.get(UpdateResult.ERRORED.name)}')
+        if dry_run:
+            logger.warning(f'The above output is the result from a dry-run pass! '
+                           f'{UpdateResult.SUCCEEDED.name} ones were not actually added! '
+                           f'{UpdateResult.ERRORED.name} is expected to be an empty list!')
+
+
+def update_email_domain(user, src_domain, dst_domain, dry_run=True):
+    """For a given user, if it has an email `example@<src_domain>`, attempt to add `example@<dst_domain>` to this
+    account. The action will be skipped if `example@<dst_domain>` already exists on this user or no email is found
+    under `@<src_domain>`. The action will fail if `example@<dst_domain>` already exists on another account. Other
+    unexpected exceptions will be considered as error.
+    """
+
+    # Find eligible emails to add
+    emails_to_add = []
+    user_result = {
+        UpdateResult.SUCCEEDED.name: [],
+        UpdateResult.SKIPPED.name: [],
+        UpdateResult.FAILED.name: [],
+        UpdateResult.ERRORED.name: [],
+    }
```

Comment:
The tricky part about taking care of multiple emails is that one may fail, another may be skipped and the rest are successful. After trying a few options, I discovered using a map is here to categorize them is the best option.